src/Components/Server/src/ComponentsHub.cs

```diff
@@ -42,32 +47,97 @@ public ComponentsHub(IServiceProvider services, ILogger<ComponentsHub> logger)
         internal CircuitHost CircuitHost
         {
             get => (CircuitHost)Context.Items[CircuitKey];
-            private set => Context.Items[CircuitKey] = value;
+            set => Context.Items[CircuitKey] = value;
         }
 
+        // For unit testing
+        internal Task OnBeforeDeactivate { get; set; } = Task.CompletedTask;
+        internal Task OnBeforeActivate { get; set; } = Task.CompletedTask;
+
         /// <summary>
         /// Intended for framework use only. Applications should not call this method directly.
         /// </summary>
         public override async Task OnDisconnectedAsync(Exception exception)
         {
-            await CircuitHost.DisposeAsync();
+            var circuitHost = CircuitHost;
+            if (circuitHost == null)
+            {
+                return;
+            }
+
+            CircuitHost = null;
+            try
+            {
+                if (!await AcquireCircuitLock())
+                {
+                    // If we're unable to enter the critical section, clear this instance and throw.
+                    await circuitHost.DisposeAsync();
+
+                    throw new InvalidOperationException("Failed to acqiure a lock to de-activate the CircuitHost instance");
+                }
+
+                await OnBeforeDeactivate;
+
+                _circuitRegistry.Deactivate(circuitHost, Context.ConnectionId);
+            }
+            finally
+            {
+                CircuitRegistryLock.Release();
+            }
+
+            await circuitHost.OnConnectionDownAsync();
         }
 
         /// <summary>
         /// Intended for framework use only. Applications should not call this method directly.
         /// </summary>
-        public async Task StartCircuit(string uriAbsolute, string baseUriAbsolute)
+        public async Task<string> StartCircuit(string uriAbsolute, string baseUriAbsolute)
         {
-            var circuitHost = _circuitFactory.CreateCircuitHost(Context.GetHttpContext(), Clients.Caller);
+            var circuitHost = _circuitFactory.CreateCircuitHost(Context, Clients.Caller);
             circuitHost.UnhandledException += CircuitHost_UnhandledException;
-
-            var uriHelper = (RemoteUriHelper)circuitHost.Services.GetRequiredService<IUriHelper>();
-            uriHelper.Initialize(uriAbsolute, baseUriAbsolute);
+            circuitHost.RemoteUriHelper.Initialize(uriAbsolute, baseUriAbsolute);
 
             // If initialization fails, this will throw. The caller will fail if they try to call into any interop API.
             await circuitHost.InitializeAsync(Context.ConnectionAborted);
 
+            _circuitRegistry.Register(circuitHost);
```

Comment:
Why doesn't this need the lock?

Reply:
The client waits for initialization to complete before it wires up reconnects. The experience on the client is a bit crummy if you disconnect during initialization (you get a `console.log`), but it's a small window of failure and we could always make it nicer if we get feedback. On the server, there isn't a risk of reconnect while this is in progress.